### PR TITLE
Don't remove completed proposals

### LINKED ISSUE #252: It is difficult to audit historical governance changes, as accepted proposals are automatically removed
When a governance proposal is successful, it is automatically removed from proposals table:

`memberfrontend.h : complete_proposal`
```
// if the vote was successful, remove the proposal
proposals->remove(id);
```

This means a member cannot discover the outcome of a proposal by querying the running service, they must examine the ledger and reconstruct the history.

The normal behaviour should be to retain all proposals, and add an enum indicating their current status (`Open`, `Accepted`, `Closed`). Where the proposer can currently unilaterally remove a proposal, they should only be able to close it to further votes.

To avoid large KV memory bloat, we may need to add a separate path to actually remove proposals. This could be only available to the proposer, or to any member, or we could require it to go through a governance process too.

Its also a sensible area to add request size limits - proposals should only be X characters at most, votes at most Y characters - and these limits are themselves subject to governance. This will lead to some very amusing governance deadlock situations, where the members make the rules so restrictive that they can't change the rules.

### PR DESCRIPTION
**Pull request by eddyashton**
Rather than instantly removing completed proposals (making governance changes harder to audit), we retain proposals with an enum indicating their current state.

New proposals are created in the `Open` state, where they can be voted for by the members. If a quorum approve, the proposal transitions to `Accepted` - the proposed changes are enacted, and the proposal will not accept further votes. The proposer may _withdraw_ an `Open` proposal, transitioning it to the `Withdrawn` state. These proposals remain visible, but will not accept further votes - they will never be `Accepted`.

I think `withdraw` describes the desired behaviour better than `reject` or similar, but open to discussion.

I've kept the existing removal behaviour, so the proposer can unilaterally remove a proposal if they wish. But in general, I think we should prefer to `withdraw` than to `remove`.

This includes some renaming:
- "removal" RPC becomes "remove", to be consistent with other verb-methods
- `Proposal` struct becomes `Propose`, to match the name of the RPC it describes
- `OpenProposal` becomes `Proposal`, since it now contains non-open Proposals as well

This resolves #252.

### CODE REVIEW

**Comment by cimetrics** [comment]
images

**Comment by cimetrics** [comment]
images

**Comment by cimetrics** [comment]
images

File: src/node/rpc/memberfrontend.h [hunk 1]
```diff
@@ -300,13 +307,50 @@ namespace ccf
...
+
+        const auto proposal_action = args.params.get<ProposalAction>();
+        const auto proposal_id = proposal_action.id;
+        auto proposals = args.tx.get_view(this->network.proposals);
+        auto proposal = proposals->get(proposal_id);
+
+        if (!proposal)
+          return jsonrpc::error(
+            jsonrpc::StandardErrorCodes::INVALID_PARAMS,
+            "Proposal does not exist");
```
Comment (olgavrou) [inline src/node/rpc/memberfrontend.h]: Should these errors return the proposal id?

Comment (eddyashton) [reply src/node/rpc/memberfrontend.h]: Yes, I'll add these

File: tests/memberclient.py [hunk 2]
```diff
@@ -55,7 +55,12 @@ def run(args):
         assert proposal_id == 0
```
Comment (olgavrou) [inline tests/memberclient.py]: Can we add tests for all the new error cases introduced? Not being able to vote on an open proposal, not being able to withdraw a proposal that isn't open, trying to withdraw someone else's proposal and failing.

Comment (eddyashton) [reply tests/memberclient.py]: I've added more tests, and I believe we cover everything now:
- Accepted votes won't process more votes
- Accepted proposal can't be withdrawn
- Non-proposers can't withdraw or remove proposal
- Withdrawn proposals can't be withdrawn again
- Withdrawn proposals won't process votes
- Removed proposals can't be withdrawn
- Removed proposals won't process votes
- Removed proposals can't be removed again

**Comment by cimetrics** [comment]
images

File: sphinx/source/governance.rst [hunk 3]
```diff
@@ -82,12 +82,16 @@ The details of pending proposals, including the proposer member ID, proposal scr
...
 
 .. code-block:: bash
 
-    $ memberclient --server-address 127.83.203.69:55526 --cert member1_cert.pem --privk member1_privk.pem --ca networkcert.pem removal --proposal-id 0
+    $ memberclient --server-address 127.83.203.69:55526 --cert member1_cert.pem --privk member1_privk.pem --ca networkcert.pem withdraw --proposal-id 0
     {"commit":110,"global_commit":109,"id":0,"jsonrpc":"2.0","result":true,"term":4}
+
+This means future votes will be ignored, and the proposal will never be accepted. However it will remain visible as a proposal so members can easily audit historic proposals.
+
+If a proposer wishes to remove a proposal entirely, so it is not retrievable from the live service, they may call the ``remove`` RPC. The proposal may still be retrieved from the ledger.
```
Comment (achamayou) [inline sphinx/source/governance.rst]: withdraw RPC?

File: sphinx/source/schemas/remove_params.json [hunk 4]
```diff
@@ -0,0 +1,15 @@
+{
```
Comment (achamayou) [inline sphinx/source/schemas/remove_params.json]: Why is there a remove and withdraw schemas? Are they different RPCs?

File: sphinx/source/governance.rst [hunk 5]
```diff
@@ -82,12 +82,16 @@ The details of pending proposals, including the proposer member ID, proposal scr
...
-At any stage during the voting process and before the proposal is completed, the proposing member may decide to remove a pending proposal:
+At any stage during the voting process and before the proposal is completed, the proposing member may decide to withdraw a pending proposal:
 
 .. code-block:: bash
 
-    $ memberclient --server-address 127.83.203.69:55526 --cert member1_cert.pem --privk member1_privk.pem --ca networkcert.pem removal --proposal-id 0
+    $ memberclient --server-address 127.83.203.69:55526 --cert member1_cert.pem --privk member1_privk.pem --ca networkcert.pem withdraw --proposal-id 0
     {"commit":110,"global_commit":109,"id":0,"jsonrpc":"2.0","result":true,"term":4}
+
+This means future votes will be ignored, and the proposal will never be accepted. However it will remain visible as a proposal so members can easily audit historic proposals.
```
Comment (achamayou) [inline sphinx/source/governance.rst]: I think it's quite confusing to have both withdraw and remove, I would prefer keeping only withdraw. The only benefit of remove seems to be a tiny bit of ledger space saving, which is likely insignificant and does not make up for the confusion in my opinion.

Comment (eddyashton) [reply sphinx/source/governance.rst]: I think I agree. I'm wary of normalising arbitrary memory growth without _some_ way of removing state, but its probably not worth adding individual removes for every operation. We should consider a generic cull/remove call in future for network management, but until its a problem we shouldn't overcomplicate the interface. Will remove.

**Comment by cimetrics** [comment]
images

**Comment by cimetrics** [comment]
images

**Pull request merged**